pkg/cmd/pulumi/stack_change_secrets_provider_test.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"os"
+	"testing"
+
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/v3/resource/stack"
+	"github.com/pulumi/pulumi/pkg/v3/secrets"
+	"github.com/pulumi/pulumi/pkg/v3/secrets/b64"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/encoding"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/config"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// Test that we validate the secrets provider and return an error
+func TestChangeSecretsProvider_Invalid(t *testing.T) {
+	t.Parallel()
+
+	var stdoutBuff bytes.Buffer
+	cmd := stackChangeSecretsProviderCmd{
+		stdout: &stdoutBuff,
+		stack:  "test",
+	}
+	err := cmd.Run(context.Background(), []string{"not_a_secret"})
+	require.Error(t, err)
+	assert.ErrorContains(t, err, "unknown secrets provider type 'not_a_secret' "+
+		"(supported values: default,passphrase,awskms,azurekeyvault,gcpkms,hashivault)")
+}
+
+func mockStdin(t *testing.T, input string) {
+	t.Helper()
+	r, w, err := os.Pipe()
+	require.NoError(t, err)
+	_, err = w.Write([]byte(input))
```

Comment:
Does `os.Pipe` buffer in memory or on-disk?
I feel like this could deadlock with a large enough `input` if it's buffered in-memory?
It might be prudent to do the write-and-close in a goroutine.

Reply:
It _could_ but given this is just for testing small passwords we shouldn't ever hit those limits.